akka-persistence/src/main/scala/akka/persistence/Persistent.scala

```diff
@@ -38,6 +38,15 @@ object AtomicWrite {
 }
 
 final case class AtomicWrite(payload: immutable.Seq[PersistentRepr]) extends PersistentEnvelope with Message {
+
+  // TODO: quite costy... do we need to do this? We only internally create AWs hm...
+  require(!payload.exists(_.persistenceId != payload.head.persistenceId),
+    s"AtomicWrite must contain messages for the same persistenceId, yet different persistenceIds found: ${payload.map(_.persistenceId).toSet}")
```

Comment:
you can optimize it by adding an if check in front of it

    // do we use List?
    if (payload.tail.nonEmpty) require(...

    // do we use Vector? (would be stupid)
    if (payload.size > 1) require(...